integration/build/build_test.go

```diff
@@ -793,6 +795,43 @@ func TestBuildEmitsImageCreateEvent(t *testing.T) {
 	}
 }
 
+func TestBuildHistoryDoesNotPreventRemoval(t *testing.T) {
+	skip.If(testEnv.DaemonInfo.OSType == "windows", "buildkit is not supported on Windows")
+
+	ctx := setupTest(t)
+
+	dockerfile := "FROM busybox\nRUN echo hello world > /hello"
+	source := fakecontext.New(t, "", fakecontext.WithDockerfile(dockerfile))
+	defer source.Close()
+
+	apiClient := testEnv.APIClient()
+
+	buildImage := func(imgName string) error {
+		resp, err := apiClient.ImageBuild(ctx, source.AsTarReader(t), types.ImageBuildOptions{
+			Remove:      true,
+			ForceRemove: true,
+			Tags:        []string{imgName},
+			Version:     types.BuilderBuildKit,
+		})
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		_, err = io.Copy(io.Discard, resp.Body)
+		return err
+	}
+
+	err := buildImage("target-a")
+	assert.NilError(t, err)
+
+	resp, err := apiClient.ImageRemove(ctx, "target-a", image.RemoveOptions{})
+	assert.NilError(t, err)
+	assert.Check(t, slices.ContainsFunc(resp, func(r image.DeleteResponse) bool {
+		return r.Deleted != ""
+	}))
+}
```

Comment:
IIRC the issue was that the image index was deleted just fine (so it would get reported in the `Deleted`) - it was only the actual blobs/snapshots that weren't GC'd due to buildkit holding the gc labels.

I think we could just make this test c8d-integration only, and then check with `ctr` if the content is indeed gone at this point.

(We probably need to also prune build cache before this?)

Reply:
I'm not sure if I tested it wrong, but my testing indicated that checking for deleted signified it was fixed: https://github.com/moby/moby/issues/46858#issuecomment-2822072127

The previous version only showed `Untagged`.

Reply:
That's also possible - if you checked and that's indeed the case then testing the content store directly is a nice improvement, but not really a blocker.

Let me know if you want to try that.